### PR TITLE
Added HttpServer class as a simple alternative to start a web server

### PR DESCRIPTION
**Pull request by bogdan-szabo-sociomantic**

### CODE REVIEW

File: src/ocean/net/http/HttpServer.d [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    /**************************************************************************
+
+        Constructor
+
+        Params:
+            finalizer = finalizer callback of the select listener
+
+    **************************************************************************/
+
+    public this ( scope FinalizeDg finalizer, HttpHandler[] handlers )
```
Comment (Geod24) [inline src/ocean/net/http/HttpServer.d]: You really don't want to use `scope` here

File: src/ocean/net/http/HttpServer.d [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        return HttpResponseCode.OK;
+    }
+
+    auto config = HttpConfig("0.0.0.0", 8080);
+    auto server = new HttpServer([ &iDoNothing, &iDoSomething ]);
+
+    // The following line registers the server with the dispatcher, which in
+    // effect makes it operational, subject to the latter's scheduling.
+    // (Commented out here to not start a server instance in a unit test.)
+    // server.start(config);
```
Comment (Geod24) [inline src/ocean/net/http/HttpServer.d]: Just use a nested function which is never called:
```d
unittest
{
    void main ()
    {
        /// Magic
    }
}
```

File: src/ocean/net/http/HttpServer.d [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+            return HttpResponseCode.BadRequest;
+        }
+
+        return HttpResponseCode.NotFound;
+    }
+
+    /**************************************************************************
+
+        Tells the request message body length.
```
Comment (Geod24) [inline src/ocean/net/http/HttpServer.d:184]: Missing newline to separate the doc

File: src/ocean/net/http/HttpServer.d [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    {
+        size_t size;
+
+        try
+        {
+            this.request.getUnsigned("Content-Length", size);
+        }
+        catch(Exception e)
+        {
+            log.error("{}", e.toString);
```
Comment (Geod24) [inline src/ocean/net/http/HttpServer.d]: `toString` allocates and walk the stack, hence it's a potential DoS vector. Usually we use `message` for this.

File: src/ocean/net/http/HttpServer.d [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            {
+                auto status = handler(this.request, response);
+
+                if ( status != HttpResponseCode.NotFound )
+                {
+                    return status;
+                }
+            }
+        }
+        catch ( Exception e )
```
Comment (Geod24) [inline src/ocean/net/http/HttpServer.d:171]: Your spacing

File: src/ocean/net/http/HttpServer.d [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+    override protected size_t request_msg_body_length ( )
+    {
+        size_t size;
+
+        try
+        {
+            this.request.getUnsigned("Content-Length", size);
+        }
+        catch(Exception e)
```
Comment (Geod24) [inline src/ocean/net/http/HttpServer.d]: confuses me

File: src/ocean/net/http/HttpServer.d [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+            [ HttpMethod.Get, HttpMethod.Put, HttpMethod.Post,
+              HttpMethod.Delete, HttpMethod.Options ]);
+        this.handlers = handlers;
+
+        this.log = Log.lookup("ocean.net.http.HttpServer.HttpRequestHandler");
+    }
+
+    /**************************************************************************
+
+        Method that handles the HTTP requests, it will call each
```
Comment (Geod24) [inline src/ocean/net/http/HttpServer.d]: I got a feeling something's missing here

File: src/ocean/net/http/HttpServer.d [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+unittest
+{
+    HttpResponseCode iDoNothing( HttpRequest, out const(char)[] )
+    {
+        return HttpResponseCode.NotFound;
+    }
+
+    HttpResponseCode iDoSomething( HttpRequest, out const(char)[] response )
+    {
+        response = "I did something".dup;
```
Comment (Geod24) [inline src/ocean/net/http/HttpServer.d]: You don't need a `dup` here

**Comment by Geod24** [comment]
So this iterates over all the handlers, in sequence ?
Anyone with a non-trivial API will want to have routes, and those are not supported as far as I can tell, and need to be handled manually, which the example unittest does not demonstrate how to do.

File: src/ocean/net/http/HttpServer.d (hunk 1 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/net/http/HttpServer.d]: weird... it's used like this everywhere in the released version. It compiles without it, so I removed `scope`

https://github.com/sociomantic-tsunami/ocean/blob/v5.1.0/integrationtest/selectlistener/UnixServer.d#L102

https://github.com/sociomantic-tsunami/ocean/blob/v5.1.0/src/ocean/net/http/HttpConnectionHandler.d#L125

https://github.com/sociomantic-tsunami/ocean/blob/v5.1.0/src/ocean/net/server/connection/IConnectionHandler.d#L161

https://github.com/sociomantic-tsunami/ocean/blob/v5.1.0/src/ocean/util/prometheus/server/PrometheusHandler.d#L66

and so on....

File: src/ocean/net/http/HttpServer.d (hunk 2 above)
Comment (bogdan-szabo-sociomantic) [reply src/ocean/net/http/HttpServer.d]: I tried to do something similar to this:

https://github.com/sociomantic-tsunami/ocean/blob/v5.1.0/src/ocean/util/prometheus/server/PrometheusListener.d#L138

which looked good enough to be merged. Thanks for the tip!

**Comment by bogdan-szabo-sociomantic** [comment]
@Geod24  thanks for the review! I did not try to write an url router here. I wrote  a web server that can handle multiple handlers. Right now the current api allows only one handler, and we can't easily use the prometheus stats with some other http handler on the same port.

**Comment by bogdan-szabo-sociomantic** [comment]
@Geod24 ping

**Comment by Geod24** [comment]
@bogdan-szabo-sociomantic : I'll leave the rest of the review to maintainers. From my PoV, an HTTP server class without routing ability is of little help over what is already existing. And it does not seem to be Ocean's goal (at least wasn't in my time) to provide those kinds of ability.

So I'd rather have someone familiar with the internal needs review this :)

**Comment by ben-palmer-sociomantic** [comment]
If you could provide a good use case here it would help.

**Comment by ben-palmer-sociomantic** [comment]
Is this still active? I will close if we are no longer working on this.

**Comment by bogdan-szabo-sociomantic** [comment]
we can close all of my branches :) we already have the code in production in our private repos

**Pull request closed without merging**